### PR TITLE
Data source missing required columns

### PR DESCRIPTION
**Pull request by kaapstorm**
Context: [FB 268655](https://manage.dimagi.com/default.asp?268655)

Report Builder was not checking that the data source included columns for all possible aggregations. When the user switched from one report type to another (e.g. Case List to Case Summary, or vice versa) it was only checking columns for indicators that were in use when the user switched. As a result, switching from a Case Summary to a Case List report would break the Case List report if the Case Summary report aggregated a numerical field. (Editing the report again (or refreshing Report Builder) and saving after an arbitrary change would recreate the data source and fix the error.)

This change checks that all required columns are included when switching report types, and if not, rebuilds the data source for all possible indicators (which is what refreshing the Report Builder would do).

(The last commit addresses the bug. The first two are refactors.)

@czue @biyeun @gbova

### CODE REVIEW

File: corehq/apps/userreports/reports/builder/forms.py [hunk 1]
```diff
@@ -419,16 +419,17 @@ def indicators(self, columns, filters, is_multiselect_chart_report=False):
                 indicator = property_.to_report_filter_indicator(filter_)
                 indicators.setdefault(str(indicator), indicator)
 
+        if as_dict:
+            return indicators
+
         return list(indicators.values())
```
Comment (czue) [inline corehq/apps/userreports/reports/builder/forms.py]: nit: might be slightly more readable if you had:
`return indicators if as_dict else list(indicators.values())`

**Review by czue — APPROVED** [review]
I'm not really familiar with this code enough to wrap my head around the fix, but everything looks good to the extent that I understood it. I'm assuming you've considered any edge cases/done any QA that is necessary to feel good rolling this out.

**Review by czue — COMMENTED** [review]
Actually just saw the build failed. Which raises a good point - is it possible/easy to write a test that illustrates the issue and fix? Seems like that would be good to have.

```
======================================================================
FAIL: corehq.apps.userreports.tests.test_report_builder:MultiselectQuestionTest.testDataSource
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_report_builder.py", line 242, in testDataSource
    self.assertEqual(len(mselect_indicators), 1)
AssertionError: 0 != 1
```

**Comment by kaapstorm** [comment]
Yeah, I agree; a test would definitely be good. I'll get on it.

File: corehq/apps/userreports/reports/builder/forms.py [hunk 2]
```diff
@@ -425,14 +425,16 @@ def indicators(self, columns, filters, is_multiselect_chart_report=False, as_dic
                     # will use one indicator for the field's string value, and "Sum" and "Average" aggregations
                     # will use a second indicator for the field's numerical value. "column_id" includes the
                     # indicator's data type, so it is unique per indicator.
-                    indicators.setdefault(indicator['column_id'], indicator)
+                    indicator_key = (indicator['column_id'], indicator['type'])
```
Comment (kaapstorm) [inline corehq/apps/userreports/reports/builder/forms.py]: @czue the failing test checked that indicators for a multi-select question included a "choice_list" indicator. The test failed because I assumed that the column_id for indicators of the same data source would be unique, but that assumption was wrong. 

These are the indicators: 
```javascript
{
  "column_id": "data_state_6e36b993", 
  "datatype": "string", 
  "display_name": "state", 
  "expression": {
    "property_path": [
      "form", 
      "state"
    ], 
    "type": "property_path"
  }, 
  "type": "expression"
}
{
  "choices": [
    "MA", 
    "MN", 
    "VT"
  ], 
  "column_id": "data_state_6e36b993", 
  "display_name": "state", 
  "property_path": [
    "form", 
    "state"
  ], 
  "select_style": "multiple", 
  "type": "choice_list"
}
```

So when this keyed the `indicators` dictionary on `indicator['column_id']`, the choice_list indicator was being replaced with the the expression indicator.

I'm worried that my assumption should be true though. Is this a bug? Should column_id be unique?

Comment (czue) [reply corehq/apps/userreports/reports/builder/forms.py]: the current validation checks that [database columns should be unique](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/models.py#L330). since choice list indicators get expanded to [one column per choice](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/userreports/indicators/factory.py#L102) it's not actually creating a conflict.

this is definitely confusing and i'd support if someone wanted to make the case that `column_id` should be unique within a data source even if it's not strictly required, but perhaps might be a bit of a detour - one potentially requiring complex data migrations - and not critical for the current work?

Comment (kaapstorm) [reply corehq/apps/userreports/reports/builder/forms.py]: Thanks. For now, at least, I'm happy with just knowing why the current behaviour is fine.
